atomdb/base.py

```diff
@@ -318,6 +424,194 @@ def __get__(self, obj: T, cls: Optional[Type[T]] = None):
+        if f in excluded:
+            continue
+        meta = m.metadata or {}
+        order = meta.get("setstate_order", 1000)
+
+        # Allow  tagging a custom unflatten fn
+        unflatten = meta.get("unflatten", default_unflatten)
+
+        setters.append((order, f, m, unflatten))
+    setters.sort(key=lambda it: it[0])
+
+    on_error = cls.__on_error__
+
+    namespace = {
+        "default_unflatten": default_unflatten,
+    }
+    for order, f, m, unflatten in setters:
+        # Since f is potentially an untrusted input, make sure it is a valid
+        # python identifier to prevent unintended code being generated.
+        if not f.isidentifier():
+            raise ValueError(f"Field '{f}' cannot be used for code generation")
+
+        template.append(f"if '{f}' in state:")
+
+        # Determine the expresion to unflatten the value
+        if unflatten is default_unflatten:
+            if is_primitive_member(m):
+                # Direct assignment
+                expr = f"state['{f}']"
+            else:
+                # Default flatten
+                expr = f"await default_unflatten(state['{f}'], scope)"
+        else:
+            namespace[f"unflatten_{f}"] = unflatten
+            if asyncio.iscoroutinefunction(unflatten):
+                expr = f"await unflatten_{f}(state['{f}'], scope)"
+            else:
+                expr = f"unflatten_{f}(state['{f}'], scope)"
+
+        # Do the assignment
+        if on_error == "raise":
+            template.append(f"    self.{f} = {expr}")
+        else:
+            if on_error == "log":
+                handler = f"self.__log_restore_error__(e, '{f}', state, scope)"
+            else:
+                handler = "pass"
+            template.extend(
+                [
+                    f"    try:",
+                    f"        self.{f} = {expr}",
+                    f"    except Exception as e:",
+                    f"        {handler}",
+                ]
+            )
+
+    # Update restored state
+    template.append("self.__restored__ = True")
+    source = "\n    ".join(template)
+    return generate_function(source, namespace, "__restorestate__")
+
+
+def generate_function(source: str, namespace: DictType[str, Any], fn_name: str):
+    """Generate an optimized function
+
+    Parameters
+    ----------
+    source: str
+        The function source code
+    namespaced: dict
+        Namespace available to the function
+    fn_name: str
+        The name of the generated function.
+
+    Returns
+    -------
+    fn: function
+        The function generated.
+
+    """
+    # print(source)
+    try:
+        assert source.startswith(f"def {fn_name}") or source.startswith(
+            f"async def {fn_name}"
+        )
+        code = compile(source, __name__, "exec", optimize=1)
+    except Exception as e:
+        raise RuntimeError(f"Could not generate code: {e}:\n{source}")
+
+    # TODO: Use byteplay to rewrite globals to load fast
```

Comment:
`rewrite_to_fast_locals` won't help here since it is used to turn expression into function body and won't be able to optimize anything here I think.

Reply:
Thanks for giving it a quick look. The author of cattrs says they got an improvement by doing something similar (see https://threeofwands.com/why-cattrs-is-so-fast/) but the benchmarks seem to show the same performance (between atomdb and cattrs) so I suspect your right.

Reply:
At least at one point in time accessing/setting atom member was faster than accessing standard Python attr (C++ indexing of the array being faster than looking into the `__dict__` hashmap, so it may be why you are already on par with it.